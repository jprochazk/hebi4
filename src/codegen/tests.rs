use std::fs::read_to_string;

use super::emit;
use crate::{parser::parse, token::tokenize};

uitest::uitest!("tests/inputs/codegen/*.hi", "codegen", |path| {
    let input = read_to_string(path).unwrap();
    let tokens = tokenize(&input);
    let ast = match parse(&tokens) {
        Ok(ast) => ast,
        Err(err) => {
            panic!("{}", err.render(&input));
        }
    };
    let (snapshot, failure) = match emit(&ast) {
        Ok(chunk) => (format!("OK\n{}", chunk.disasm(&input)), false),
        Err(err) => (format!("ERROR\n{}", err.render(&input)), true),
    };

    #[cfg(not(miri))]
    {
        insta::assert_snapshot!(snapshot);
        let _ = failure;
    }

    #[cfg(miri)]
    {
        if failure {
            panic!("{snapshot}");
        } else {
            eprintln!("{snapshot}");
        }
    }
});
