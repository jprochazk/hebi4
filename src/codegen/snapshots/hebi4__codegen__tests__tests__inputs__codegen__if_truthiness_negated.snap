---
source: src/codegen/tests.rs
expression: snapshot
---
OK
.fn @main
  params 0
  stack 4
  .locals
    x = r1
    y = r2
  .code
   0  lsmi r1, 1
   1  lsmi r2, 0
   2  isfalse r1
   3  jmp 3   ; to 6
   4  lsmi r0, 1
   5  ret
   6  isfalse r2
   7  jmp 4   ; to 11
   8  lsmi r0, 2
   9  ret
  10  lnil r0
  11  ret
